actionpack/lib/action_controller/metal/content_security.rb

```diff
@@ -0,0 +1,308 @@
+        build_directives(context).compact.join("; ") + ";"
+      end
+
+      private
+        def copy_directives(directives)
+          {}.tap do |copy|
+            directives.each do |directive, sources|
+              copy[directive] = sources.map(&:dup)
+            end
+          end
+        end
+
+        def apply_mappings(sources)
+          sources.map do |source|
+            case source
+            when Symbol
+              apply_mapping(source)
+            when String, Proc
+              source
+            else
+              raise ArgumentError, "Invalid content security policy source: #{source.inspect}"
+            end
+          end
+        end
+
+        def apply_mapping(source)
+          if MAPPINGS.key?(source)
+            MAPPINGS.fetch(source)
+          else
+            raise ArgumentError, "Unknown content security policy source mapping: #{source.inspect}"
+          end
+        end
+
+        def build_directives(context)
+          @directives.map do |directive, sources|
+            if sources.is_a?(Array)
+              "#{directive} #{build_directive(sources, context).join(' ')}"
+            elsif sources
+              directive
+            else
+              nil
+            end
+          end
+        end
+
+        def build_directive(sources, context)
+          sources.map { |source| resolve_source(source, context) }
+        end
+
+        def resolve_source(source, context)
+          case source
+          when String
+            source
+          when Symbol
+            source.to_s
+          when Proc
+            if context.nil?
+              raise RuntimeError, "Missing context for the dynamic content security policy source: #{source.inspect}"
+            else
+              context.instance_exec(&source)
+            end
+          else
+            raise RuntimeError, "Unexpected content security policy source: #{source.inspect}"
+          end
+        end
+    end
+
+    included do
+      config_accessor :content_security_policies
+      self.content_security_policies = PolicyCollection.new
+
+      config_accessor :default_protect_content
+      self.default_protect_content = false
+
+      attr_accessor :content_security_policy
+    end
+
+    module ClassMethods
+      # TODO: Documentation for `protect_content`
+      def protect_content(options = {}, &block)
+        policy_name = options.delete(:policy)
+
+        before_action(options) do
+          if policy_name
+            policy = content_security_policies.fetch(policy_name)
+          end
+
+          if block_given?
+            if policy
+              policy = policy.clone
+            else
+              policy = Policy.new
+            end
+
+            yield policy
+          end
+
+          unless policy
+            raise RuntimeError, "Please specify a content security policy"
+          end
+
+          unless policy.empty?
+            self.content_security_policy = policy
+          end
```

Comment:
Maybe an edge case, but it could be sensible to set the header even if the policy is empty, making it clear that we've set an empty policy rather than omitting it.

Reply:
Yes, agreed.